src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -246,36 +283,55 @@ ray::Status OwnershipBasedObjectDirectory::UnsubscribeObjectLocations(
+    // call, and we have received at least one update from the owner about
+    // the object's creation, then call the callback immediately with the
+    // cached locations.
+    auto &locations = it->second.current_object_locations;
+    auto &spilled_url = it->second.spilled_url;
+    auto &spilled_node_id = it->second.spilled_node_id;
+    auto object_size = it->second.object_size;
+    io_service_.post(
+        [callback, object_id, locations, spilled_url, spilled_node_id, object_size]() {
+          callback(object_id, locations, spilled_url, spilled_node_id, object_size);
+        });
+  } else {
+    WorkerID worker_id = WorkerID::FromBinary(owner_address.worker_id());
+    std::shared_ptr<rpc::CoreWorkerClient> rpc_client = GetClient(owner_address);
+    if (rpc_client == nullptr) {
+      RAY_LOG(WARNING) << "Object " << object_id << " does not have owner. "
+                       << "LookupLocations returns an empty list of locations.";
+      io_service_.post([callback, object_id]() {
+        callback(object_id, std::unordered_set<NodeID>(), "", NodeID::Nil(), 0);
+      });
+      return Status::OK();
+    }
 
-  rpc::GetObjectLocationsOwnerRequest request;
-  request.set_intended_worker_id(owner_address.worker_id());
-  request.set_object_id(object_id.Binary());
-  request.set_last_version(-1);
+    rpc::GetObjectLocationsOwnerRequest request;
+    request.set_intended_worker_id(owner_address.worker_id());
+    request.set_object_id(object_id.Binary());
+    request.set_last_version(-1);
 
-  rpc_client->GetObjectLocationsOwner(
-      request, [this, worker_id, object_id, callback](
-                   Status status, const rpc::GetObjectLocationsOwnerReply &reply) {
-        if (!status.ok()) {
-          RAY_LOG(ERROR) << "Worker " << worker_id << " failed to get the location for "
-                         << object_id;
-        }
-        std::unordered_set<NodeID> node_ids;
-        for (auto const &node_id : reply.node_ids()) {
-          node_ids.emplace(NodeID::FromBinary(node_id));
-        }
-        FilterRemovedNodes(gcs_client_, &node_ids);
-        callback(object_id, node_ids, "", NodeID::Nil(), reply.object_size());
-      });
+    rpc_client->GetObjectLocationsOwner(
+        request, [this, worker_id, object_id, callback](
+                     Status status, const rpc::GetObjectLocationsOwnerReply &reply) {
+          if (!status.ok()) {
+            RAY_LOG(ERROR) << "Worker " << worker_id << " failed to get the location for "
+                           << object_id;
+          }
+          std::unordered_set<NodeID> node_ids;
+          std::string spilled_url;
+          NodeID spilled_node_id;
+          size_t object_size = 0;
+          UpdateObjectLocations(reply, status, object_id, gcs_client_, &node_ids,
+                                &spilled_url, &spilled_node_id, &object_size);
+          // It is safe to call the callback directly since this is already running
```

Comment:
Isn't this always safe? We only run them in a single thread

Reply:
It's not safe due to [this issue](https://github.com/ray-project/ray/issues/2959), further described in [this thread](https://github.com/ray-project/ray/pull/3227#discussion_r230614532).

Reply:
I think the comment is a bit confusing. Why don’t we add an issue link to explain what it means by “safe”? (Or just remove the whole comment)

Reply:
Sure, I'll add comments near all of these considerations. Btw, this is the same comment [as in the current GCS-based object directory](https://github.com/ray-project/ray/blob/9a6a3b40c1731e214e88155befd6126fa8ac4eb4/src/ray/object_manager/object_directory.cc#L282-L283).